journal/journal.go

```diff
@@ -50,19 +50,35 @@ const (
 	PriDebug
 )
 
-var conn net.Conn
+var (
+	conn *net.UnixConn
+	// This can be overridden at build-time:
+	// https://github.com/golang/go/wiki/GcToolchainTricks#including-build-information-in-the-executable
+	journalSocket = "/run/systemd/journal/socket"
+)
 
 func init() {
-	var err error
-	conn, err = net.Dial("unixgram", "/run/systemd/journal/socket")
+	autobind, err := net.ResolveUnixAddr("unixgram", "")
+	if err != nil {
+		return
+	}
+	conn, err = net.ListenUnixgram("unixgram", autobind)
 	if err != nil {
 		conn = nil
+		return
 	}
 }
 
 // Enabled returns true if the local systemd journal is available for logging
 func Enabled() bool {
-	return conn != nil
+	if conn == nil {
```

Comment:
admittedly, this is unrelated to this PR, but I wonder if `init()` has any happens-before guarantees in the runtime.

If `Enabled()` is being invoked in a different goroutine other than the main goroutine which invoked `init()` this would be an unsynchronized read, potentially returning any arbitrary value.

Reply:
nevermind, `init()` _is_ part of https://golang.org/ref/spec#Package_initialization, so happens-before should apply here :+1:

Reply:
That's a fair concern though. While goroutines can't race here, other packages' `init` functions calling `journal.Enabled` could be scheduled before our `init`. If that is a concern, possible solutions would be:
 * avoid global variables (which is tempting, but moves socket creation in the hot path of log emitting)
 * wrap `conn` initialization in a `sync.Once` and use it here too

Reply:
`sync.Once` sounds like a good solution :+1: